### DISCUSSION TITLE
[Custom construtor] How to bind a member function as a cumstom construtor?

### DISCUSSION THREAD

**Discussion by huangweiwu** [wjakob/nanobind#266] [Q&A] [answered]
I have a existing class A with a member function create which return a shared_ptr<A> object.
How to bind create as a cumtom construtor?
In Boost.Python, this can be achieved through the following approach.
.def("__init__", make_constructor(&A::create))

**Comment by wjakob** [comment, marked as answer]
You need to move the instance ourself:

```cpp
.def("__init__", [](A* a) { new (a) (A::create()); });
```

**Reply by huangweiwu** [reply]
What if the create function is a template function and returns a shared_ptr<T>?
Should I bind a nb::init<>() function instead of def("__init__") since we don't need to specify holder types in nanobind?
